mongo/change_stream.go

```diff
@@ -143,6 +143,21 @@ func newChangeStream(ctx context.Context, config changeStreamConfig, pipeline in
 	if cs.options.MaxAwaitTime != nil {
 		cs.cursorOptions.MaxTimeMS = int64(*cs.options.MaxAwaitTime / time.Millisecond)
 	}
+	if cs.options.CustomOptions != nil {
+		// Marshal all custom options before passing to the initial aggregate. Return
+		// any errors from Marshaling.
+		customOptions := make(map[string]bsoncore.Value)
+		for optionName, optionValue := range cs.options.CustomOptions {
+			bsonType, bsonData, err := bson.MarshalValueWithRegistry(cs.registry, optionValue)
+			if err != nil {
+				cs.err = err
```

Comment:
I think this is necessary to prevent leaking the implicit session. It is done in other error returns in this function.

```suggestion
				cs.err = err
				closeImplicitSession(cs.sess)
```

Reply:
Added the missing `closeImplicitSession()`. 

Using that deferred error check is a little tricky in this case. Upon error, the `newChangeStream` function tries to both return the error and set the `ChangeStream`'s error (`cs.err`). I'd have to start using a new `err` named return on top of `cs.err` that could be checked in the deferred block. Note that checking `cs.err` in the deferred block usually won't work because `newChangeStream` has called `return nil, cs.Err()` and set the named `cs` return back to `nil`. All very confusing; it's odd that we try to set `cs.err` _and_ return it, but I don't think the deferred implicit session close will be as clean here.

Reply:
Sounds good, thanks for the detailed explanation!